File: examples/Go/tasksink2.go
// Task design 2
// Adds pub-sub flow to send kill signal to workers
//
// Author:  Brendan Mc.
// Requires: http://github.com/alecthomas/gozmq

package main

import (
    "fmt"
    zmq "github.com/alecthomas/gozmq"
    "time"
)

func main() {
	context, _ := zmq.NewContext()
	defer context.Close()
    
    // Socket to receive messages on
    receiver, _ := context.NewSocket(zmq.PULL)
    defer receiver.Close()
    receiver.Bind("tcp://*:5558")
    
    // Socket for worker control
    controller, _ := context.NewSocket(zmq.PUB)
    defer controller.Close()
    controller.Bind("tcp://*:5559")
    
    // Wait for start of batch
    _, _ = receiver.Recv(0)
    
    // Process 100 confirmations
    tstart := time.Now()
    for i := 0;i < 100;i = i + 1 {
        _, _ = receiver.Recv(0)
        
        if i % 10 == 0 {
            fmt.Print(":")
        } else {
            fmt.Print(".")
        }
    }
    
    total := time.Since(tstart).Nanoseconds() / 1000000
    fmt.Printf("\nTotal elapsed time:  %d msec\n", total)
    
    controller.Send([]byte("KILL"), 0)
    return
}

File: examples/PHP/tasksink2.php
<?php
/*
 *  Task design 2
 *  Adds pub-sub flow to send kill signal to workers
 * @author Ian Barber <ian(dot)barber(at)gmail(dot)com>
 */

$context = new ZMQContext();

//  Socket to receive messages on
$receiver = new ZMQSocket($context, ZMQ::SOCKET_PULL);
$receiver->bind("tcp://*:5558");

//  Socket for worker control
$controller = new ZMQSocket($context, ZMQ::SOCKET_PUB);
$controller->bind("tcp://*:5559");

//  Wait for start of batch
$string = $receiver->recv();

//  Process 100 confirmations
$tstart = microtime(true);
$total_msec = 0;     //  Total calculated cost in msecs
for ($task_nbr = 0; $task_nbr < 100; $task_nbr++) {
	$string = $receiver->recv();
    
	if($task_nbr % 10 == 0) {
		echo ":";
	} else {
		echo ".";
	}
}

$tend = microtime(true);

$total_msec = ($tend - $tstart) * 1000;
echo PHP_EOL;
printf ("Total elapsed time: %d msec", $total_msec);
echo PHP_EOL;

//  Send kill signal to workers
$controller->send("KILL");

//  Finished
sleep (1);              //  Give 0MQ time to deliver
